KAFKA-3355

mgarmes opened the pull request:
KAFKA-3355
GetOffsetShell command doesn't work with SASL enabled Kafka

mgarmes committed:
Merge remote-tracking branch 'refs/remotes/apache/trunk'

mgarmes committed:
KAFKA-3355

GetOffsetShell command doesn't work with SASL enabled Kafka

ijuma commented:
Thanks for the PR. Is there a reason to keep the old implementation?

mgarmes commented:
Hello,
 
Can someone check this pull request  ?

best regard

ijuma commented:
@garmes-gdev, did you see my question?

mgarmes commented:
Hello @ijuma,
Thank you very much for your reply.
regarding you question, for my personal opinion I don't want to keep the old version, but I noticed in the others PR, contributors keep the both version so I did the same. if you think is it not useful I can delete it !! 

best regards.

ijuma commented:
@garmes-gdev, the old version only needs to be kept if there's a reason to do so (maybe the new version is missing some functionality or is not as stable). If there is no reason to keep the old version, then we can simply replace it.

Commit:
KAFKA-3355 GetOffsetShell command doesn't work with SASL enabled.

mgarmes commented:
Hello @ijuma, 
I eliminated the old version and I updated the tests.

Aegeaner committed:
KAFKA-4709：Error message from Struct.validate() should include the name of the offending field.

https://issues.apache.org/jira/browse/KAFKA-4709

Author: Aegeaner <xihuke@gmail.com>

Reviewers: Dong Lin, Guozhang Wang

Closes #2521 from Aegeaner/KAFKA-4709

hachikuji committed:
MINOR: Move compression stream construction into CompressionType

Author: Jason Gustafson <jason@confluent.io>

Reviewers: Ismael Juma <ismael@juma.me.uk>

Closes #2536 from hachikuji/minor-move-compression-io-construction

dguy committed:
KAFKA-4484: Set more conservative default values on RocksDB for memory usage

Lowered the default RocksDB settings for the block cache and write buffers

Author: Damian Guy <damian.guy@gmail.com>

Reviewers: Eno Thereska, Guozhang Wang

Closes #2525 from dguy/kafka-4484

cmccabe committed:
KAFKA-4775; Fix findbugs warnings in kafka-tools

Author: Colin P. Mccabe <cmccabe@confluent.io>

Reviewers: Ismael Juma <ismael@juma.me.uk>

Closes #2559 from cmccabe/KAFKA-4775

becketqin committed:
KAFKA-4340; Follow-up fixing system test failures and handling non default log.retention.ms

Author: Jiangjie Qin <becket.qin@gmail.com>

Reviewers: Jun Rao <junrao@gmail.com>, Ismael Juma <ismael@juma.me.uk>

Closes #2544 from becketqin/KAFKA-4340_follow_up

ijuma committed:
MINOR: Remove Struct from Request/Response classes

More details:
* Replaced `struct` field in Request/Response with a `toStruct` method. This
makes the performance model (including memory usage) easier to understand.
Note that requests have `toStruct()` while responses have `toStruct(version)`.
* Replaced mutable `version` field in `Request.Builder` with an immutable
field `desiredVersion` and a `version` parameter passed to the `build` method.
* Optimised `handleFetchRequest` to avoid unnecessary creation of `Struct`
instances (from 4 to 2 in the worst case and 2 to 1 in the best case).
* Various clean-ups in request/response classes and their test. In particular,
it is now clear what we are testing. Previously, it looked like we were testing
more than we really were.

With this in place, we could remove `AbstractRequest.Builder` in the future by
doing the following:
* Change `AbstractRequest.toStruct` to accept a version (like responses).
* Change `AbstractRequest.version` to be `desiredVersion` (like `Builder`).
* Change `ClientRequest` to take `AbstractRequest`.
* Move validation from the `build` methods to the request constructors or
static factory methods.
* Anything else required for the code to compile again.

Author: Ismael Juma <ismael@juma.me.uk>

Reviewers: Apurva Mehta <apurva.1618@gmail.com>, Jason Gustafson <jason@confluent.io>

Closes #2513 from ijuma/separate-struct

dguy committed:
KAFKA-4317: Regularly checkpoint StateStore changelog offsets

Currently the checkpoint file is deleted at state store initialization and it is only ever written again during a clean shutdown. This can result in significant delays during restarts as the entire store needs to be loaded from the changelog.
We can mitigate against this by frequently checkpointing the offsets. The checkpointing happens only during the commit phase, i.e, after we have manually flushed the store and the producer. So we guarantee that the checkpointed offsets are never greater than what has been flushed.
In the event of hard failure we can recover by reading the checkpoints and consuming from the stored offsets.

Author: Damian Guy <damian.guy@gmail.com>

Reviewers: Eno Thereska, Matthias J. Sax, Guozhang Wang

Closes #2471 from dguy/kafka-4317

hachikuji committed:
HOTFIX: ClassCastException in request logging

Author: Jason Gustafson <jason@confluent.io>

Reviewers: Colin P. Mccabe <cmccabe@confluent.io>, Ewen Cheslack-Postava <me@ewencp.org>

Closes #2566 from hachikuji/hotfix-request-logging

Commit:
MINOR: Replace for with foreach loop in common module

Author: Prabhat Kashyap <prabhat.kashyap@knoldus.in>

Reviewers: Ismael Juma <ismael@juma.me.uk>

Closes #2530 from PKOfficial/refactored-code

Commit:
MINOR: Increase consumer init timeout in throttling test

The throttling system test sometimes fail because it takes longer than the current 10 second time out for partitions to get assigned to the consumer.

Author: Apurva Mehta <apurva@confluent.io>

Reviewers: Ismael Juma <ismael@juma.me.uk>

Closes #2567 from apurvam/increase-timeout-for-partitions-assigned

cmccabe committed:
KAFKA-4774; Inner classes which don't need a reference to the outer c…

…lass should be static

Author: Colin P. Mccabe <cmccabe@confluent.io>

Reviewers: Jason Gustafson <jason@confluent.io>, Ismael Juma <ismael@juma.me.uk>

Closes #2558 from cmccabe/KAFKA-4774

Commit:
KAFKA-4777; Backoff properly in consumer heartbeat thread if no brokers are available

Author: Allen Xiang <allen.xiang@monsanto.com>

Reviewers: Jason Gustafson <jason@confluent.io>

Closes #2564 from allenxiang/client-heartbeat-fix

original-brownbear committed:
HOTFIX: ClassCastException in Request error logging

Fixed ClassCastException resulting from missing type hint in request logging.

Author: Armin Braun <me@obrown.io>

Reviewers: Jason Gustafson <jason@confluent.io>

Closes #2571 from original-brownbear/fix-logging-err-response

original-brownbear committed:
MINOR: Fixed 3 inner classes without instance reference to be static

* Turned 3 inner classes that weren't static but could be into `static` ones.
* Turned one `public` inner class that wasn't used publicly into a `private`.

Trivial but imo worthwhile to explicitly keep visibility and "staticness" correct in syntax (if only to be nice to the GC) :)

Author: Armin Braun <me@obrown.io>

Reviewers: Matthias J. Sax <matthias@confluent.io>, Jason Gustafson <jason@confluent.io>

Closes #2574 from original-brownbear/cleanup-inner-nonstatic

hachikuji committed:
KAFKA-4776; Implement graceful handling for improperly formed compressed message sets

Author: Jason Gustafson <jason@confluent.io>

Reviewers: Jiangjie Qin <becket.qin@gmail.com>

Closes #2572 from hachikuji/KAFKA-4776

hachikuji committed:
MINOR: Fix NPE handling unknown APIs in NodeApiVersions.toString

Author: Jason Gustafson <jason@confluent.io>

Reviewers: Colin P. Mccabe <cmccabe@confluent.io>, Ismael Juma <ismael@juma.me.uk>

Closes #2561 from hachikuji/fix-npe-api-version-tostring

cshannon committed:
MINOR: Add build_eclipse to .gitignore

build_eclipse is the configured output directory for eclipse when using
the gradle eclipse plugin and should be ignored

Author: Christopher L. Shannon <christopher.l.shannon@gmail.com>

Reviewers: Ismael Juma <ismael@juma.me.uk>

Closes #2569 from cshannon/eclipse-gitignore

hachikuji committed:
MINOR: Remove unused MessageWriter and CompressionFactory

Author: Jason Gustafson <jason@confluent.io>

Reviewers: Apurva Mehta <apurva@confluent.io>, Ismael Juma <ismael@juma.me.uk>

Closes #2543 from hachikuji/remove-message-writer

Commit:
KAFKA-4752: Fixed bandwidth calculation

Author: Eno Thereska <eno@confluent.io>

Reviewers: Damian Guy, Guozhang Wang

Closes #2551 from enothereska/KAFKA-4752-join-bw

cmccabe committed:
KAFKA-4757; NetworkClient should log request details at trace level when a request is cancelled because of disconnection

Author: Colin P. Mccabe <cmccabe@confluent.io>

Reviewers: Onur Karaman <okaraman@linkedin.com>, Apurva Mehta <apurva.1618@gmail.com>, Jiangjie Qin <becket.qin@gmail.com>

Closes #2563 from cmccabe/KAFKA-4757

cmccabe committed:
KAFKA-4708; Fix transient failure in BrokerApiVersionsCommandTest.checkBrokerApiVersionCommandOutput

Author: Colin P. Mccabe <cmccabe@confluent.io>

Reviewers: Jason Gustafson <jason@confluent.io>, Dong Lin <lindong28@gmail.com>, Ismael Juma <ismael@juma.me.uk>

Closes #2489 from cmccabe/KAFKA-4708

rajinisivaram committed:
KAFKA-4786; Wait for heartbeat thread to terminate in consumer close

Author: Rajini Sivaram <rajinisivaram@googlemail.com>

Reviewers: Apurva Mehta <apurva@confluent.io>, Ismael Juma <ismael@juma.me.uk>, Jason Gustafson <jason@confluent.io>

Closes #2586 from rajinisivaram/KAFKA-4786

ijuma committed:
MINOR: Move ProtoUtils methods to ApiKeys

Also move `requireTimestamp` to `minVersion` logic from `Fetcher` to
`ListOffsetRequest.Builder.forConsumer()`.

Author: Ismael Juma <ismael@juma.me.uk>

Reviewers: Colin P. Mccabe <cmccabe@confluent.io>, Jason Gustafson <jason@confluent.io>

Closes #2580 from ijuma/move-proto-utils-to-api-keys

ijuma committed:
KAFKA-4788: Revert "KAFKA-4092: retention.bytes should not be allowed to be less than segment.bytes"

The intent is good, but it needs to take into account broker configs as well.
See KAFKA-4788 for more details.

This reverts commit 4ca5abe8ee7578f602fb7653cb8a09640607ea85.

Author: Ismael Juma <ismael@juma.me.uk>

Reviewers: Jun Rao <junrao@gmail.com>

Closes #2588 from ijuma/kafka-4788

enothereska committed:
MINOR: Reduce stream thread metrics overhead

Reduces overheads by avoiding re-calling time.milliseconds().

Author: Eno Thereska <eno.thereska@gmail.com>

Reviewers: Guozhang Wang <wangguoz@gmail.com>

Closes #2583 from enothereska/minor-reduce-milliseconds

guozhangwang committed:
HOTFIX: add js template to upgrade.html

Author: Guozhang Wang <wangguoz@gmail.com>

Reviewers: Ismael Juma <ismael@juma.me.uk>

Closes #2591 from guozhangwang/KHotfix-explicit-version

original-brownbear committed:
KAFKA-4198; Fix race condition in KafkaServer.shutdown()

It contained this step:

      val canShutdown = isShuttingDown.compareAndSet(false, true)
      if (canShutdown && shutdownLatch.getCount > 0) {

without any fallback for the case of `shutdownLatch.getCount == 0`. So in the case
of `shutdownLatch.getCount == 0` (when a previous call to the shutdown method
was right about to finish) you would set `isShuttingDown` to true again without any
possibility of ever getting the server started (since `startup` will check
`isShuttingDown` before setting up a new latch with count 1).

Long story short: concurrent calls to shutdown can get the server locked in a broken state.

This fixes the reported error:

java.lang.IllegalStateException: Kafka server is still shutting down, cannot re-start!
	at kafka.server.KafkaServer.startup(KafkaServer.scala:184)
	at kafka.integration.KafkaServerTestHarness$$anonfun$restartDeadBrokers$2.apply$mcVI$sp(KafkaServerTestHarness.scala:117)
	at kafka.integration.KafkaServerTestHarness$$anonfun$restartDeadBrokers$2.apply(KafkaServerTestHarness.scala:116)
	at kafka.integration.KafkaServerTestHarness$$anonfun$restartDeadBrokers$2.apply(KafkaServerTestHarness.scala:116)
	at scala.collection.TraversableLike$WithFilter$$anonfun$foreach$1.apply(TraversableLike.scala:733)
	at scala.collection.immutable.Range.foreach(Range.scala:160)
	at scala.collection.TraversableLike$WithFilter.foreach(TraversableLike.scala:732)
	at kafka.integration.KafkaServerTestHarness$class.restartDeadBrokers(KafkaServerTestHarness.scala:116)
	at kafka.api.ConsumerBounceTest.restartDeadBrokers(ConsumerBounceTest.scala:34)
	at kafka.api.ConsumerBounceTest$BounceBrokerScheduler.doWork(ConsumerBounceTest.scala:158)

Author: Armin Braun <me@obrown.io>

Reviewers: Ismael Juma <ismael@juma.me.uk>

Closes #2568 from original-brownbear/KAFKA-4198

kamalcph committed:
MINOR: Code quality improvements to Config classes

- Minor Javadoc fixes
- Used final modifier if possible
- Unnecessary type casts removed
- Other minor clean-ups

Author: Kamal C <kamal.chandraprakash@gmail.com>

Reviewers: Colin P. Mccabe <cmccabe@confluent.io>, Ismael Juma <ismael@juma.me.uk>

Closes #2578 from Kamal15/config

jpdaigle committed:
MINOR: fixed javadoc typo in KafkaProducer.partitionsFor

Author: Jean-Philippe Daigle <jdaigle@tripadvisor.com>

Reviewers: Ismael Juma <ismael@juma.me.uk>

Closes #2592 from jpdaigle/minor_typo_fix

rajinisivaram committed:
KAFKA-4779; Fix security upgrade system test to be non-disruptive

The phase_two security upgrade test verifies upgrading inter-broker and client protocols to the same value as well as different values. The second case currently changes inter-broker protocol without first enabling the protocol, disrupting produce/consume until the whole cluster is updated. This commit changes the test to be a non-disruptive upgrade test that enables protocols first (simulating phase one of upgrade).

Author: Rajini Sivaram <rajinisivaram@googlemail.com>

Reviewers: Apurva Mehta <apurva.1618@gmail.com>, Ismael Juma <ismael@juma.me.uk>

Closes #2589 from rajinisivaram/KAFKA-4779

dguy committed:
KAFKA-4494: Reduce startup and rebalance time by batching restoration of stores

Replace one-by-one initialization of state stores with bulk initialization.

Author: Damian Guy <damian.guy@gmail.com>

Reviewers: Eno Thereska, Guozhang Wang

Closes #2560 from dguy/kafka-4494

ijuma committed:
MINOR: Fix transient failure of testCannotSendToInternalTopic

It’s a simple matter of creating the internal topic before trying to send
to it. Otherwise, we could get an `UnknownTopicOrPartitionException`
in some cases.

Without the change, I could reproduce a failure in less than
5 runs. With the change, 30 consecutive runs succeeded.

Author: Ismael Juma <ismael@juma.me.uk>

Reviewers: Apurva Mehta <apurva.1618@gmail.com>, Jason Gustafson <jason@confluent.io>

Closes #2584 from ijuma/test-cannot-send-to-internal-topic-transient-failure

original-brownbear committed:
MINOR: Fixed Non-Final Close Method + its Duplication

Author: Armin Braun <me@obrown.io>

Reviewers: Ismael Juma <ismael@juma.me.uk>, Jason Gustafson <jason@confluent.io>

Closes #2582 from original-brownbear/cleanup-nonfinal-close

ijuma committed:
MINOR: Make it impossible to invoke `Request.body` without an explicit type parameter

Author: Ismael Juma <ismael@juma.me.uk>

Reviewers: Jason Gustafson <jason@confluent.io>

Closes #2579 from ijuma/safer-body

hachikuji committed:
MINOR: Ensure consumer calls poll() if requests are outstanding

Author: Jason Gustafson <jason@confluent.io>

Reviewers: Onur Karaman <okaraman@linkedin.com>, Ismael Juma <ismael@juma.me.uk>

Closes #2596 from hachikuji/ensure-poll-with-inflight-requests

ijuma committed:
MINOR: Fix potential integer overflow and String.format issue

Author: Ismael Juma <ismael@juma.me.uk>

Reviewers: Apurva Mehta <apurva.1618@gmail.com>, Jason Gustafson <jason@confluent.io>

Closes #2585 from ijuma/overflow-and-format-fixes

vahidhashemian committed:
MINOR: Use API hyperlinks in 'Kafka Protocol Guide' to facilitate navigation

Finding the protocol associated with an API key can be a challenge in the lengthy [web page](http://kafka.apache.org/protocol.html#protocol_api_keys).
Adding hyperlinks would definitely help with that.

Co-authored with imandhan.

Author: Vahid Hashemian <vahidhashemian@us.ibm.com>

Reviewers: Ewen Cheslack-Postava <ewen@confluent.io>

Closes #2467 from vahidhashemian/minor/hyperlinks_in_kafka_protocol_guide

Gacko committed:
KAFKA-4806; Prevent double logging of ConsumerConfig

The consumer properties get logged twice since two instances
of ConsumerConfig are created during creation of KafkaConsumer.

I added a constructor of ConsumerConfig accepting the boolean
parameter doLog which is already passable in AbstractConfig
and set it to false during the second ConsumerConfig creating
in the KafkaConsumer constructor.

Author: Marco Ebert <marco_ebert@icloud.com>

Reviewers: Ismael Juma <ismael@juma.me.uk>

Closes #2600 from Gacko/trunk

enothereska committed:
KAFKA-4744: Increased timeout for bounce test

Author: Eno Thereska <eno.thereska@gmail.com>

Reviewers: Ismael Juma, Matthias J. Sax, Guozhang Wang

Closes #2601 from enothereska/KAFKA-4744-bounce

C0urante committed:
MINOR: Make asJsonSchema() and asConnectSchema() methods public

Want to use these methods in an external project.

Author: Chris Egerton <fearthecellos@gmail.com>

Reviewers: Ewen Cheslack-Postava <ewen@confluent.io>

Closes #2610 from C0urante/public-json-schema-conversion

cmccabe committed:
KAFKA-4809: docker/run_tests.sh should set up /opt/kafka-dev to be the source directory

…0.x and not 0.8

Author: Colin P. Mccabe <cmccabe@confluent.io>

Reviewers: Ewen Cheslack-Postava <ewen@confluent.io>

Closes #2602 from cmccabe/KAFKA-4809

mjsax committed:
MINOR: improve license header check by providing head file instead of (prefix) header regex

Author: Matthias J. Sax <matthias@confluent.io>

Reviewers: Jason Gustafson <jason@confluent.io>, Ewen Cheslack-Postava <ewen@confluent.io>

Closes #2303 from mjsax/licenseHeader

Commit:
MINOR: Fix typo in javadoc of `flatMapValues`

Author: Michael G. Noll <michael@confluent.io>

Reviewers: Damian Guy, Matthias J. Sax, Guozhang Wang

Closes #2607 from miguno/trunk-flatMapValues-docstring

hrafzali committed:
KAFKA-4789: Added support to ProcessorTopologyTestDriver to forward timestamps to internal topics

This resolves the issue in the ProcessorTopologyTestDriver that the extracted timestamp is not forwarded with the produced record to the internal topics.

JIRA ticket: https://issues.apache.org/jira/browse/KAFKA-4789

The contribution is my original work and I license the work to the project under the project's open source license.

guozhangwang dguy

Author: Hamidreza Afzali <hrafzali@gmail.com>

Reviewers: Damian Guy, Guozhang Wang

Closes #2590 from hrafzali/KAFKA-4789_ProcessorTopologyTestDriver_timestamp

dguy committed:
MINOR: Add code quality checks (and suppressions) to checkstyle.xml

Author: Damian Guy <damian.guy@gmail.com>

Reviewers: Guozhang Wang <wangguoz@gmail.com>, Ewen Cheslack-Postava <me@ewencp.org>, Jason Gustafson <jason@confluent.io>, Ismael Juma <ismael@juma.me.uk>

Closes #2594 from dguy/checkstyle

cmccabe committed:
KAFKA-4773; The Kafka build should run findbugs

Author: Colin P. Mccabe <cmccabe@confluent.io>

Reviewers: Ismael Juma <ismael@juma.me.uk>

Closes #2557 from cmccabe/KAFKA-4773

dguy committed:
KAFKA-4677 Follow Up: add optimization to StickyTaskAssignor for rolling rebounce

Detect when a rebalance has happened due to one or more existing nodes bouncing. Keep assignment of previous active tasks the same and only assign the tasks that were not active to the new clients.

Author: Damian Guy <damian.guy@gmail.com>

Reviewers: Guozhang Wang <wangguoz@gmail.com>

Closes #2609 from dguy/kstreams-575

dguy committed:
MINOR: improve MinTimestampTrackerTest and fix NPE when null element removed

Author: Damian Guy <damian.guy@gmail.com>

Reviewers: Matthias J. Sax, Guozhang Wang

Closes #2611 from dguy/testing

enothereska committed:
MINOR: Minor reduce unnecessary calls to time.millisecond (part 2)

Avoid calling time.milliseconds more often than necessary. Cleaning and committing logic can use the timestamp at the start of the loop with minimal consequences. 5-10% improvements noticed with request rates of 450K records/second.

Also tidy up benchmark code a bit more.

Author: Eno Thereska <eno.thereska@gmail.com>
Author: Eno Thereska <eno@confluent.io>

Reviewers: Matthias J. Sax, Damian Guy, Guozhang Wang

Closes #2603 from enothereska/minor-reduce-milliseconds2

huxihx committed:
KAFKA4811; ReplicaFetchThread may fail to create due to existing metric

Have fetcherThreadMap keyed off brokerId + fetcherId instead of broker + fetcherId, but did not consider the case where port is changed.

Author: huxi <huxi@zhenrongbao.com>

Reviewers: Jun Rao <junrao@gmail.com>

Closes #2606 from amethystic/kafka4811_ReplicaFetchThread_fail_create

sslavic committed:
KAFKA-2358: Cluster collection returning methods never return null

See https://issues.apache.org/jira/browse/KAFKA-2358

Author: Stevo Slavic <sslavic@gmail.com>

Reviewers: Jason Gustafson, Guozhang Wang

Closes #96 from sslavic/feature/KAFKA-2358

bbejeck committed:
KAFKA-4652: Added tests for KStreamBuilder

Author: bbejeck <bbejeck@gmail.com>

Reviewers: Damian Guy, Guozhang Wang

Closes #2597 from bbejeck/KAFKA-4652_improve_kstream_builder_test_coverage

Commit:
MINOR: Fix typo in GlobalKTable javadocs

Author: Michael G. Noll <michael@confluent.io>

Reviewers: Matthias J. Sax, Damian Guy, Guozhang Wang

Closes #2626 from miguno/trunk-globalktable-typos

Commit:
KAFKA-4800: Streams State transition ASCII diagrams need fixing and polishing

added \<pre> tags to not break javadoc display of the ASCII diagrams.
see broken ascii here:
https://kafka.apache.org/0102/javadoc/org/apache/kafka/streams/KafkaStreams.State.html

fix can be checked with gradle :streams:javadoc and then checking streams/build/docs/javadoc/org/apache/kafka/streams/KafkaStreams.State.html

I also fixed the diagram in StreamThread.java however currently no javadoc is generated for that one (since it's internal)

enothereska please have a look

Author: Clemens Valiente <clemens.valiente@trivago.com>

Reviewers: Eno Thereska, Matthias J. Sax, Guozhang Wang

Closes #2621 from cvaliente/KAFKA-4800-ASCII-diagrams

rajinisivaram committed:
KAFKA-4631; Request metadata in consumer if topic/partitions unavailable

If leader node of one more more partitions in a consumer subscription are temporarily unavailable, request metadata refresh so that partitions skipped for assignment dont have to wait for metadata expiry before reassignment. Metadata refresh is also requested if a subscribe topic or assigned partition doesn't exist.

Author: Rajini Sivaram <rajinisivaram@googlemail.com>

Reviewers: Vahid Hashemian <vahidhashemian@us.ibm.com>, Ismael Juma <ismael@juma.me.uk>, Jason Gustafson <jason@confluent.io>

Closes #2622 from rajinisivaram/KAFKA-4631

vahidhashemian committed:
KAFKA-2857; Retry querying the consumer group while initializing

This applies to new-consumer based groups and would avoid scenarios in which user issues a `--describe` query while the group is initializing.
Example: The following could occur for a newly created group.
```
kafkakafka:~/workspace/kafka$ bin/kafka-consumer-groups.sh --bootstrap-server localhost:9092 --describe --group g
Note: This will only show information about consumers that use the Java consumer API (non-ZooKeeper-based consumers).

Error: Executing consumer group command failed due to The group coordinator is not available.
```

With this PR the group is queried repeatedly at specific intervals within a preset (and configurable) timeout `group-init-timeout` to circumvent unfortunate situations like above.

Author: Vahid Hashemian <vahidhashemian@us.ibm.com>

Reviewers: Jason Gustafson <jason@confluent.io>

Closes #2538 from vahidhashemian/KAFKA-2857

ijuma committed:
MINOR: Bump version to 0.11.0.0-SNAPSHOT

There won't be a 0.10.3.0.

Author: Ismael Juma <ismael@juma.me.uk>

Reviewers: Jason Gustafson <jason@confluent.io>

Closes #2628 from ijuma/bump-version-to-0.11.0.0-SNAPSHOT

cmccabe committed:
MINOR: Fix tests/docker/Dockerfile

Fix tests/docker/Dockerfile to put the old Kafka distributions in the
correct spot for tests.  Also, run_tests.sh should exit with an error
code if image rebuilding fails, rather than silently falling back to an
older image.

Author: Colin P. Mccabe <cmccabe@confluent.io>

Reviewers: Ewen Cheslack-Postava <ewen@confluent.io>

Closes #2613 from cmccabe/dockerfix

vahidhashemian committed:
KAFKA-2857; MINOR: Follow up to
MINOR: Fix ResetIntegrationTest test failures

KAFKA-2857 follow-up.

Author: Vahid Hashemian <vahidhashemian@us.ibm.com>

Reviewers: Jason Gustafson <jason@confluent.io>, Ismael Juma <ismael@juma.me.uk>

Closes #2636 from vahidhashemian/minor/kafka-2857-followup

cmccabe committed:
KAFKA-4796; Fix some findbugs warnings in Kafka Java client

Author: Colin P. Mccabe <cmccabe@confluent.io>

Reviewers: Ismael Juma <ismael@juma.me.uk>

Closes #2593 from cmccabe/KAFKA-4796

original-brownbear committed:
KAFKA-3182; Fix testSocketsCloseOnShutdown transient failures

* Turned off Nagle on the sending sockets to force the socket to physically acknowledge after the first write in `sendRequest`
* Added a `200ms` delay between write attempts (necessary on Linux, but not Mac)

Author: Armin Braun <armin.braun@1und1.de>

Reviewers: Ismael Juma <ismael@juma.me.uk>

Closes #2632 from original-brownbear/KAFKA-3182

lindong28 committed:
KAFKA-4820; ConsumerNetworkClient.send() should not require global lock

Author: Dong Lin <lindong28@gmail.com>

Reviewers: Jason Gustafson <jason@confluent.io>, Ismael Juma <ismael@juma.me.uk>, Jiangjie Qin <becket.qin@gmail.com>

Closes #2619 from lindong28/KAFKA-4820

Commit:
MINOR: Standardised benchmark params for consumer and streams

There were some minor differences in the basic consumer config and streams config that are now rectified. In addition, in AWS environments the socket size makes a big difference to performance and I've tuned it up accordingly. I've also increased the number of records now that perf is higher.

Author: Eno Thereska <eno@confluent.io>

Reviewers: Guozhang Wang <wangguoz@gmail.com>

Closes #2634 from enothereska/minor-standardize-params

cmccabe committed:
KAFKA-4826: Fix some findbugs warnings in Kafka Streams

Author: Colin P. Mccabe <cmccabe@confluent.io>

Reviewers: Matthias J. Sax, Guozhang Wang

Closes #2623 from cmccabe/KAFKA-4826

Commit:
KAFKA-4276: Add REST configuration in connector properties

Addition of REST configuration in connect-distributed.properties config file
gwenshap ewencp - Please review.

Author: Akhilesh Naidu <akhilesh_naidu@persistent.com>

Reviewers: Gwen Shapira

Closes #2505 from akhilesh1194/JIRA_KAFKA-4276

soenkeliebau committed:
KAFKA-4567 - Connect Producer and Consumer ignore ssl parameters…

KAFKA-4567 - Connect Producer and Consumer ignore ssl parameters configured for worker

Added brief explanation to the docs about parameter inheritance of Kafka consumers and producers from the worker config.

Author: Sönke Liebau <soenke.liebau@opencore.com>

Reviewers: Gwen Shapira

Closes #2511 from soenkeliebau/KAFKA-4567

hrafzali committed:
KAFKA-4828: ProcessorTopologyTestDriver does not work when using through

This resolves the following issues in the ProcessorTopologyTestDriver:

- It should not create an internal changelog topic when using `through()` and `table()`
- It should forward the produced record back into the topology if it is to a source topic

Jira ticket: https://issues.apache.org/jira/browse/KAFKA-4828

The contribution is my original work and I license the work to the project under the project's open source license.

Author: Hamidreza Afzali <hrafzali@gmail.com>

Reviewers: Matthias J. Sax, Guozhang Wang

Closes #2629 from hrafzali/KAFKA-4828_ProcessorTopologyTestDriver_through

vahidhashemian committed:
MINOR: additional refactoring around the use of Errors

A couple of updates were missed in the [PR](https://github.com/apache/kafka/pull/2475) that replaced the use of error codes with Errors objects.

Author: Vahid Hashemian <vahidhashemian@us.ibm.com>

Reviewers: Ismael Juma <ismael@juma.me.uk>

Closes #2635 from vahidhashemian/minor/Errors_refactoring_leftover

bbejeck committed:
KAFKA-3989; Initial support for adding a JMH benchmarking module

Author: bbejeck <bbejeck@gmail.com>

Reviewers: Ewen Cheslack-Postava <ewen@confluent.io>, Ismael Juma <ismael@juma.me.uk>

Closes #1712 from bbejeck/KAFKA-3989_create_jmh_benchmarking_module

benstopford committed:
KAFKA-4266; ReassignPartitionsClusterTest: ensure ZK publication is completed before start

Increase the reliability of the one temporal comparison in ReassignPartitionsClusterTest by imposing a delay after ZK is updated. This should be more reliable than just increasing the amount of data.

This relates to a previous PR: https://github.com/apache/kafka/pull/1982

Author: Ben Stopford <benstopford@gmail.com>

Reviewers: Ismael Juma <ismael@juma.me.uk>

Closes #1997 from benstopford/KAFKA-4266

mjsax committed:
HOTFIX: fix broken link for wordcount demo example

Author: Matthias J. Sax <matthias@confluent.io>

Reviewers: Guozhang Wang <wangguoz@gmail.com>

Closes #2644 from mjsax/hotfixBrokerLink

original-brownbear committed:
KAFKA-3155; Avoid long overflow in RecordBatch#maybeExpire

`Long.MaxValue` for the linger overflows in `RecordBatch#maybeExpire` when added to
the current timestamp.

Then causes an error to be set for the batch by `Sender` (not happening every time since
it depends on the timing of `Sender`):

That error then causes a call to `ProduceRequestResult#done` on the batch, which then
makes the check for "not done" fail.

Author: Armin Braun <me@obrown.io>

Reviewers: Ismael Juma <ismael@juma.me.uk>

Closes #2639 from original-brownbear/KAFKA-3155

hachikuji committed:
MINOR: Rename RecordBatch to ProducerBatch to free the name for KIP-98

Author: Jason Gustafson <jason@confluent.io>

Reviewers: Apurva Mehta <apurva.1618@gmail.com>, Ismael Juma <ismael@juma.me.uk>

Closes #2646 from hachikuji/rename-record-batch

dguy committed:
KAFKA-4851: only search available segments during Segments.segments(from, to)

restrict the locating of segments in `Segments#segments(..)` to only the segments that are currently available, i.e., rather than searching the hashmap for many segments that don't exist.

Author: Damian Guy <damian.guy@gmail.com>

Reviewers: Guozhang Wang <wangguoz@gmail.com>

Closes #2645 from dguy/session-windows-testing

lindong28 committed:
KAFKA-4841; NetworkClient should only consider a connection to have failed after attempt to connect

Also fix a potential reordering bug and include a few clean-ups.

Author: Dong Lin <lindong28@gmail.com>

Reviewers: Jiangjie (Becket) Qin <becket.qin@gmail.com>, Jason Gustafson <jason@confluent.io>, Ismael Juma <ismael@juma.me.uk>

Closes #2641 from lindong28/KAFKA-4820-followup

sharad-develop committed:
KAFKA-4722: Add application.id to StreamThread name

Add application.id to StreamThread name

Author: sharad.develop <sharad.develop@gmail.com>

Reviewers: Matthias J. Sax, Guozhang Wang

Closes #2617 from sharad-develop/KAFKA-4722

hachikuji committed:
MINOR: Add varint serde utilities for new message format

Author: Jason Gustafson <jason@confluent.io>

Reviewers: Ismael Juma <ismael@juma.me.uk>

Closes #2647 from hachikuji/add-varint-serdes

hachikuji committed:
MINOR: Use ConcurrentMap for ConsumerNetworkClient UnsentRequests

Author: Jason Gustafson <jason@confluent.io>

Reviewers: Ismael Juma <ismael@juma.me.uk>

Closes #2656 from hachikuji/minor-cleanup-unsent-requests

Commit:
KAFKA-4745; Remove unnecessary flush in KafkaLZ4BlockOutputStream.close()

Remove unnecessary 'flush', the underlying stream should handle it on close.

Author: Will Droste <william.droste@arris.com>

Reviewers: Ismael Juma <ismael@juma.me.uk>

Closes #2528 from wdroste/trunk

simplesteph committed:
KAFKA-4864; added correct zookeeper nodes for security migrator

Author: simplesteph <stephane.maarek@gmail.com>

Reviewers: Jun Rao <junrao@gmail.com>

Closes #2655 from simplesteph/fix-security-migrator-tool

ijuma committed:
MINOR: Introduce NetworkClient.hasInFlightRequests

It’s a minor optimisation, but simple enough.

Author: Ismael Juma <ismael@juma.me.uk>

Reviewers: Apurva Mehta <apurva.1618@gmail.com>, Jason Gustafson <jason@confluent.io>

Closes #2658 from ijuma/has-in-flight-requests

raghavgautam committed:
KAFKA-4467: Run tests on travis-ci using docker

ijuma ewencp cmccabe harshach Please review.
Here is a sample run:
https://travis-ci.org/raghavgautam/kafka/builds/191714520

In this run 214 tests were run and 144 tests passed.

I will open separate jiras for fixing failures.

Author: Raghav Kumar Gautam <raghav@apache.org>

Reviewers: Sriharsha Chintalapani <harsha@hortonworks.com>, Ewen Cheslack-Postava <ewen@confluent.io>

Closes #2376 from raghavgautam/trunk

ijuma committed:
KAFKA-4861; GroupMetadataManager record is rejected if broker configured with LogAppendTime

The record should be created with CreateTime (like in the producer). The conversion to
LogAppendTime happens automatically (if necessary).

Author: Ismael Juma <ismael@juma.me.uk>

Reviewers: Jason Gustafson <jason@confluent.io>

Closes #2657 from ijuma/kafka-4861-log-append-time-breaks-group-data-manager

Commit:
MINOR: KStream: fix typo in javadoc

Author: Michael G. Noll <michael@confluent.io>

Reviewers: Guozhang Wang, Damian Guy

Closes #2668 from miguno/trunk-kstream-javadoc-typo

sharad-develop committed:
KAFKA-4738: Remove generic type of class ClientState

Remove generic type of class ClientState and generic T inTaskAssignor.

Author: sharad.develop <sharad.develop@gmail.com>

Reviewers: Matthias J. Sax, Guozhang Wang

Closes #2616 from sharad-develop/KAFKA-4738

Commit:
MINOR: increase RocksDb parallelism

Author: Eno Thereska <eno@confluent.io>
Author: Eno Thereska <eno.thereska@gmail.com>

Reviewers: Damian Guy, Guozhang Wang

Closes #2663 from enothereska/minor-rocksdb-parallel

acoburn committed:
MINOR: incorrect javadoc formatting

When wrapped in a `{code ...}` block, `&lt;&gt;` are not formatted as `<>`.

For instance, see the formatting of the example in https://kafka.apache.org/0102/javadoc/org/apache/kafka/streams/KafkaStreams.html

Author: Aaron Coburn <acoburn@amherst.edu>

Reviewers: Matthias J. Sax, Guozhang Wang

Closes #2671 from acoburn/javadoc_formatting

Commit:
KAFKA-4769: Add Float serializer, deserializer, serde

Author: Michael G. Noll <michael@confluent.io>

Reviewers: Dongjin Lee, Eno Thereska, Damian Guy, Colin P. McCabe, Matthas J. Sax, Guozhang Wang

Closes #2554 from miguno/KAFKA-4769

adyach committed:
KAFKA-4657: Improve test coverage of CompositeReadOnlyWindowStore

This commmit brings improved test coverage for window store fetch method
and WindowStoreIterator

Author: Andrey Dyachkov <andrey.dyachkov@zalando.de>

Reviewers: Damian Guy, Guozhang Wang

Closes #2672 from adyach/trunk

onurkaraman committed:
KAFKA-4891; kafka.request.logger TRACE regression

Both the headers and requests have regressed to just show object ids instead of their contents from their underlying structs. I'm guessing this regression came from commit [fc1cfe475e8ae8458d8ddf119ce18d0c64653a70](https://github.com/apache/kafka/commit/fc1cfe475e8ae8458d8ddf119ce18d0c64653a70)

Author: Onur Karaman <okaraman@linkedin.com>

Reviewers: Jiangjie Qin <becket.qin@gmail.com>

Closes #2678 from onurkaraman/KAFKA-4891

guozhangwang committed:
HOTFIX: Fix apache headers in float serde class files

hachikuji

Author: Guozhang Wang <wangguoz@gmail.com>

Reviewers: Jason Gustafson <jason@confluent.io>

Closes #2679 from guozhangwang/KHotfix-serde-headers

dguy committed:
KAFKA-4863; Querying window store may return unwanted keys

Make sure that the iterator returned from `WindowStore.fetch(..)` only returns matching keys, rather than all keys that are a prefix match.

Author: Damian Guy <damian.guy@gmail.com>

Reviewers: Eno Thereska, Guozhang Wang

Closes #2662 from dguy/kafka-4863

ewencp committed:
KAFKA-4783: Add ByteArrayConverter (KIP-128)

Author: Ewen Cheslack-Postava <me@ewencp.org>

Reviewers: Guozhang Wang <wangguoz@gmail.com>

Closes #2599 from ewencp/kafka-4783-byte-array-converter

mgarmes committed:
KAFKA 3355 update

ijuma committed:
HOTFIX: Fix header in ByteArrayConverter

Author: Ismael Juma <ismael@juma.me.uk>

Reviewers: Manikumar reddy O <manikumar.reddy@gmail.com>

Closes #2690 from ijuma/fix-header-in-byte-array-converter

guozhangwang committed:
MINOR: Improve log4j on stream thread and stream process

Author: Guozhang Wang <wangguoz@gmail.com>

Reviewers: Matthias J. Sax <matthias@confluent.io>, Eno Thereska <eno@confluent.io>, Damian Guy <damian.guy@gmail.com>, Jason Gustafson <jason@confluent.io>

Closes #2685 from guozhangwang/KMinor-improve-log4j

vahidhashemian committed:
MINOR: Fix a documentation typo

Author: Vahid Hashemian <vahidhashemian@us.ibm.com>

Reviewers: Sriharsha Chintalapani <harsha@hortonworks.com>

Closes #2674 from vahidhashemian/minor/fix_typos_1703

mgarmes committed:
fix and tests (status: PASS)

mgarmes committed:
rebase

mgarmes committed:
Merge branch 'master' of https://github.com/garmes-gdev/kafka

mgarmes committed:
fix

ppanero commented:
Hello, is there any intention to merge/work more on this? @ijuma 
Thanks in advance

mgarmes commented:
Hello @ppanero,
a provide the fix for kafka 10 and is not merged so is no more compatible with 10.2 ! so I will try to provide the fix for this new version and I hop this time that someone merge it 

Bests

ppanero commented:
Thanks @garmes-gdev  let me know if u need any help and I could contribute :D

mgarmes commented:
@ppanero  https://github.com/apache/kafka/pull/2891
